include/boost/gil/image_processing/threshold.hpp

```diff
@@ -308,6 +318,153 @@ void threshold_optimal
     }
 }
 
+namespace detail {
+
+template
+<
+    typename SourceChannelT,
+    typename ResultChannelT,
+    typename SrcView,
+    typename DstView,
+    typename Operator
+>
+void adaptive_impl
+(
+    SrcView const& src_view,
+    SrcView const& convolved_view,
+    DstView const& dst_view,
+    Operator const& threshold_op
+)
+{
+    //template argument validation
+    gil_function_requires<ImageViewConcept<SrcView>>();
+    gil_function_requires<MutableImageViewConcept<DstView>>();
+    gil_function_requires<ColorSpacesCompatibleConcept
+        <
+        typename color_space_type<SrcView>::type,
+        typename color_space_type<DstView>::type>
+    >();
+
+    //iterate over the image chaecking each pixel value for the threshold
+    for (std::ptrdiff_t y = 0; y < src_view.height(); y++)
+    {
+        typename SrcView::x_iterator src_it = src_view.row_begin(y);
+        typename SrcView::x_iterator convolved_it = convolved_view.row_begin(y);
+        typename DstView::x_iterator dst_it = dst_view.row_begin(y);
+
+        for (std::ptrdiff_t x = 0; x < src_view.width(); x++)
+        {
+            static_transform(src_it[x], convolved_it[x], dst_it[x], threshold_op);
+        }
+    }
+}
+
+template <typename SrcView, typename DstView, typename DataType>
+void convolution2d(
+    SrcView src_view, DstView dst_view, std::vector<DataType> kernel, std::size_t center
+)
+{
+    int flip_ker_ind, flip_ker_col, row_boundary, col_boundary;
+    for (std::ptrdiff_t view_row = 0; view_row < src_view.height(); ++view_row)              // rows
+    {
+        for (std::ptrdiff_t view_col = 0; view_col < src_view.width(); ++view_col)          // columns
+        {
+            for (std::size_t kernel_row = 0; kernel_row < kernel.size(); ++kernel_row)     // kernel rows
+            {
+                flip_ker_ind = kernel.size() - 1 - kernel_row;      // row index of flipped kernel
+
+                for (std::size_t kernel_col = 0; kernel_col < kernel.size(); ++kernel_col) // kernel columns
+                {
+                    flip_ker_col = kernel.size() - 1 - kernel_col;  // column index of flipped kernel
+
+                    // index of input signal, used for checking boundary
+                    row_boundary = view_row + (center - flip_ker_ind);
+                    col_boundary = view_col + (center - flip_ker_col);
+
+                    // ignore input samples which are out of bound
+                    if (row_boundary >= 0 && row_boundary < src_view.height() &&
+                        col_boundary >= 0 && col_boundary < src_view.width())
+                    {
+                        dst_view(view_row, view_col) =
```

Comment:
@miralshah365  First, let's address the run-time failure you mentioned on Gitter.

Please, add these run-time checks before the `dst_view(view_row, view_col)` assignment:

```
BOOST_ASSERT(row_boundary < src_view.height());
BOOST_ASSERT(col_boundary < src_view.height());
BOOST_ASSERT(view_row < dst_view.height());
BOOST_ASSERT(view_col < dst_view.width());
```

AFAICT, indices calculation is incorrect and leads to writing memory beyond valid range which eventually leads to heap corruption (of the image memory) leading to run-time failures you are observing.

Reply:
@miralshah365 As you have spotted already, I made mistake in one of the assertions above
`BOOST_ASSERT(col_boundary < src_view.height());` 
should read
BOOST_ASSERT(col_boundary < src_view.width());

However, image heap corruption must be due to a bug somewhere else.

Reply:
@miralshah365 Well, it seems indices of sources and destination view are used as mismatched:

```
for (std::ptrdiff_t view_row = 0; view_row < src_view.height(); ++view_row)
```

then

```
dst_view(view_row, view_col)
```

where `view_row` is in range of `src_view`, isn't it?

Reply:
@mloskot but in the function I have passes one copy of `src_view`. look at line 432 of `threshold.hpp`

Reply:
@miralshah365 Compare this addressing here
```
dst_view(view_row, view_col)
```
and
```
src_view(row_boundary, col_boundary)
```

with this meaning of the arguments:

 https://github.com/boostorg/gil/blob/74c497c36fe00d5597702e7d1349848cf9c25a67/include/boost/gil/image_view.hpp#L159